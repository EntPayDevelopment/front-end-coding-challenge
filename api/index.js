import { mock } from "./mock";

export const getComedies = () => {
  return new Promise((resolve, _) => {
    setTimeout(() => {
      resolve(mock);
    }, 300);
  });
};
