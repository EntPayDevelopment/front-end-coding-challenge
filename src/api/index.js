import { mock } from "./mock";

export const getMedias = () => {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      resolve(mock);
    }, 300);
  });
};

export const addToWatchlist = (id) => {
  const successRate = 0.95;
  const success = Math.random() <= successRate ? true : false;
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      if (success) {
        resolve({ message: `Added media ${id}`, status: 200 });
      } else {
        reject({ message: `Error ${id}`, status: 422 });
      }
    }, 300);
  });
};

export const removeFromWatchlist = (id) => {
  return new Promise((resolve, _) => {
    setTimeout(() => {
      resolve({ message: `Removed media ${id}`, status: 200 });
    }, 300);
  });
};
