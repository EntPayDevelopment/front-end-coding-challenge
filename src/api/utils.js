import { toast } from 'react-toastify';

export const handleApi = async (apiCallBack) => {
    try {
        const result = await apiCallBack();
        if (result.message) {
            toast.success(result.message)
        }
        return result;
    } catch (error) {
        handleError(error);
    }
}

const handleError = (error) => {
    switch (error.status) {
        case 422:
            toast.error(error.message);
            break;
        default:
            if (error.message) {
                toast.error(error.message)
            } else {
                toast.error('Server error')
            }
            break;
    }
}