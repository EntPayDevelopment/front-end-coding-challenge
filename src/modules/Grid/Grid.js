import React from "react";
import { GridItem } from './GridItem/GridItem';
import styles from './Grid.module.css';

export const Grid = (props) => {
    const items = props.items || [];
    const gridItems = items.map((item) => {
        return (
            <GridItem key={item.id} item={item} />
        );
    });

    return (
        <section className={styles.Grid}>
            {gridItems}
        </section>
    );
};

