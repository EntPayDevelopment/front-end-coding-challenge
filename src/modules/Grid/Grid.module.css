.Grid{ 
    color: white; 
    padding: 20px;
    display: flex;
    max-width: 100%;
    flex-wrap: wrap; 
}