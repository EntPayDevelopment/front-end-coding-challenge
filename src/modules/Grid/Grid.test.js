import renderer from 'react-test-renderer';
import { Grid } from './Grid';
import { mock } from '../../api/mock';

it("[Given] No Items [Then] it will render correctly.", () => {
    const grid = renderer.create(<Grid />);
    let tree = grid.toJSON();
    expect(tree).toMatchSnapshot();
});

it("[Given] Valid Items [Then] it will render correctly.", () => {
    const grid = renderer.create(<Grid items={mock} />);
    let tree = grid.toJSON();
    expect(tree).toMatchSnapshot();
});

