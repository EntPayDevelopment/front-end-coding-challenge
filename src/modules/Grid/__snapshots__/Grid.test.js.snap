// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Given] No Items [Then] it will render correctly. 1`] = `
<section
  className="Grid"
/>
`;

exports[`[Given] Valid Items [Then] it will render correctly. 1`] = `
<section
  className="Grid"
>
  <div
    className="GridItem"
  >
    <img
      alt="poster for Our Flag Means Death"
      className="Poster"
      src="https://images2.9c9media.com/image_asset/2022_2_25_25b23590-2669-48f5-8768-7a740d8c5744_jpg_2000x3000.jpg"
    />
    <div
      className="Title"
    >
      Our Flag Means Death
    </div>
    <button
      className="Btn Add"
      data-testid="WatchlistButton"
      onClick={[Function]}
    >
      (+) Add to Watchlist
    </button>
  </div>
  <div
    className="GridItem"
  >
    <img
      alt="poster for Shining Vale"
      className="Poster"
      src="https://images2.9c9media.com/image_asset/2022_2_14_59abaa4c-7af8-4243-bec0-839bb475a65c_png_2000x3000.jpg"
    />
    <div
      className="Title"
    >
      Shining Vale
    </div>
    <button
      className="Btn Add"
      data-testid="WatchlistButton"
      onClick={[Function]}
    >
      (+) Add to Watchlist
    </button>
  </div>
  <div
    className="GridItem"
  >
    <img
      alt="poster for Friends"
      className="Poster"
      src="https://images2.9c9media.com/image_asset/2022_3_9_91c9ffc5-b3fa-4780-a776-de6989ca25ae_png_2000x3000.jpg"
    />
    <div
      className="Title"
    >
      Friends
    </div>
    <button
      className="Btn Add"
      data-testid="WatchlistButton"
      onClick={[Function]}
    >
      (+) Add to Watchlist
    </button>
  </div>
  <div
    className="GridItem"
  >
    <img
      alt="poster for Letterkenny: International Women's Day"
      className="Poster"
      src="https://images2.9c9media.com/image_asset/2022_3_8_0e28de35-baed-4928-b133-a4fcc77c6112_jpg_2000x3000.jpg"
    />
    <div
      className="Title"
    >
      Letterkenny: International Women's Day
    </div>
    <button
      className="Btn Add"
      data-testid="WatchlistButton"
      onClick={[Function]}
    >
      (+) Add to Watchlist
    </button>
  </div>
  <div
    className="GridItem"
  >
    <img
      alt="poster for Winning Time: The Rise of the Lakers Dynasty"
      className="Poster"
      src="https://images2.9c9media.com/image_asset/2022_3_2_6a9434e5-58bd-46c3-b3ab-1410cf401e40_jpg_2000x3000.jpg"
    />
    <div
      className="Title"
    >
      Winning Time: The Rise of the Lakers Dynasty
    </div>
    <button
      className="Btn Add"
      data-testid="WatchlistButton"
      onClick={[Function]}
    >
      (+) Add to Watchlist
    </button>
  </div>
  <div
    className="GridItem"
  >
    <img
      alt="poster for Hacks"
      className="Poster"
      src="https://images2.9c9media.com/image_asset/2021_8_26_982cd5e5-a7fc-461f-8fe9-e7ba33986482_png_2000x3000.jpg"
    />
    <div
      className="Title"
    >
      Hacks
    </div>
    <button
      className="Btn Add"
      data-testid="WatchlistButton"
      onClick={[Function]}
    >
      (+) Add to Watchlist
    </button>
  </div>
  <div
    className="GridItem"
  >
    <img
      alt="poster for South Park"
      className="Poster"
      src="https://images2.9c9media.com/image_asset/2019_6_25_0912d848-4ae0-482f-8841-0b24b1b1220c_png_2000x3000.jpg"
    />
    <div
      className="Title"
    >
      South Park
    </div>
    <button
      className="Btn Add"
      data-testid="WatchlistButton"
      onClick={[Function]}
    >
      (+) Add to Watchlist
    </button>
  </div>
  <div
    className="GridItem"
  >
    <img
      alt="poster for Dollface"
      className="Poster"
      src="https://images2.9c9media.com/image_asset/2022_2_8_539766e6-966d-45a0-8ef3-7487f0c096e6_png_2000x3000.jpg"
    />
    <div
      className="Title"
    >
      Dollface
    </div>
    <button
      className="Btn Add"
      data-testid="WatchlistButton"
      onClick={[Function]}
    >
      (+) Add to Watchlist
    </button>
  </div>
  <div
    className="GridItem"
  >
    <img
      alt="poster for Letterkenny"
      className="Poster"
      src="https://images2.9c9media.com/image_asset/2021_12_17_64f49892-adc2-4ccb-b19d-495eb550098d_png_2000x3000.jpg"
    />
    <div
      className="Title"
    >
      Letterkenny
    </div>
    <button
      className="Btn Add"
      data-testid="WatchlistButton"
      onClick={[Function]}
    >
      (+) Add to Watchlist
    </button>
  </div>
  <div
    className="GridItem"
  >
    <img
      alt="poster for Peacemaker"
      className="Poster"
      src="https://images2.9c9media.com/image_asset/2022_1_10_d5dfcca2-be82-40a5-93e3-dd25c94ba118_png_2000x3000.jpg"
    />
    <div
      className="Title"
    >
      Peacemaker
    </div>
    <button
      className="Btn Add"
      data-testid="WatchlistButton"
      onClick={[Function]}
    >
      (+) Add to Watchlist
    </button>
  </div>
  <div
    className="GridItem"
  >
    <img
      alt="poster for Last Week Tonight With John Oliver"
      className="Poster"
      src="https://images2.9c9media.com/image_asset/2022_2_16_31dddb6c-9c6f-4516-8556-97af6527cc8d_png_2000x3000.jpg"
    />
    <div
      className="Title"
    >
      Last Week Tonight With John Oliver
    </div>
    <button
      className="Btn Add"
      data-testid="WatchlistButton"
      onClick={[Function]}
    >
      (+) Add to Watchlist
    </button>
  </div>
  <div
    className="GridItem"
  >
    <img
      alt="poster for RuPaul's Drag Race"
      className="Poster"
      src="https://images2.9c9media.com/image_asset/2022_1_4_6761c172-9654-4de5-bd55-103abe01bcb9_png_2000x3000.jpg"
    />
    <div
      className="Title"
    >
      RuPaul's Drag Race
    </div>
    <button
      className="Btn Add"
      data-testid="WatchlistButton"
      onClick={[Function]}
    >
      (+) Add to Watchlist
    </button>
  </div>
  <div
    className="GridItem"
  >
    <img
      alt="poster for The Righteous Gemstones"
      className="Poster"
      src="https://images2.9c9media.com/image_asset/2022_1_6_0ebdcdc0-89db-4bb0-99e6-6db2b793f78b_png_2000x3000.jpg"
    />
    <div
      className="Title"
    >
      The Righteous Gemstones
    </div>
    <button
      className="Btn Add"
      data-testid="WatchlistButton"
      onClick={[Function]}
    >
      (+) Add to Watchlist
    </button>
  </div>
  <div
    className="GridItem"
  >
    <img
      alt="poster for Children Ruin Everything"
      className="Poster"
      src="https://images2.9c9media.com/image_asset/2022_2_21_3e517210-17ab-4f2a-959c-47d8259ac07f_png_2000x3000.jpg"
    />
    <div
      className="Title"
    >
      Children Ruin Everything
    </div>
    <button
      className="Btn Add"
      data-testid="WatchlistButton"
      onClick={[Function]}
    >
      (+) Add to Watchlist
    </button>
  </div>
  <div
    className="GridItem"
  >
    <img
      alt="poster for Ride the Eagle"
      className="Poster"
      src="https://images2.9c9media.com/image_asset/2022_2_11_5a588b08-ea79-468d-851c-3caf422e8f03_png_2000x3000.jpg"
    />
    <div
      className="Title"
    >
      Ride the Eagle
    </div>
    <button
      className="Btn Add"
      data-testid="WatchlistButton"
      onClick={[Function]}
    >
      (+) Add to Watchlist
    </button>
  </div>
  <div
    className="GridItem"
  >
    <img
      alt="poster for Somebody Somewhere"
      className="Poster"
      src="https://images2.9c9media.com/image_asset/2022_1_10_93be06fb-35d9-4264-8641-c82150b19c4e_png_2000x3000.jpg"
    />
    <div
      className="Title"
    >
      Somebody Somewhere
    </div>
    <button
      className="Btn Add"
      data-testid="WatchlistButton"
      onClick={[Function]}
    >
      (+) Add to Watchlist
    </button>
  </div>
  <div
    className="GridItem"
  >
    <img
      alt="poster for The Fresh Prince of Bel-Air"
      className="Poster"
      src="https://images2.9c9media.com/image_asset/2020_11_9_e69d1277-e0ba-4786-92b2-f2184f3a6975_png_2000x3000.jpg"
    />
    <div
      className="Title"
    >
      The Fresh Prince of Bel-Air
    </div>
    <button
      className="Btn Add"
      data-testid="WatchlistButton"
      onClick={[Function]}
    >
      (+) Add to Watchlist
    </button>
  </div>
  <div
    className="GridItem"
  >
    <img
      alt="poster for Astrid and Lilly Save the World"
      className="Poster"
      src="https://images2.9c9media.com/image_asset/2022_1_24_af845b5f-f280-4145-a9eb-99a208d50cec_png_2000x3000.jpg"
    />
    <div
      className="Title"
    >
      Astrid and Lilly Save the World
    </div>
    <button
      className="Btn Add"
      data-testid="WatchlistButton"
      onClick={[Function]}
    >
      (+) Add to Watchlist
    </button>
  </div>
  <div
    className="GridItem"
  >
    <img
      alt="poster for The White Lotus"
      className="Poster"
      src="https://images2.9c9media.com/image_asset/2021_8_26_643d2668-36d8-4f19-8b06-22a92904cfc1_png_2000x3000.jpg"
    />
    <div
      className="Title"
    >
      The White Lotus
    </div>
    <button
      className="Btn Add"
      data-testid="WatchlistButton"
      onClick={[Function]}
    >
      (+) Add to Watchlist
    </button>
  </div>
  <div
    className="GridItem"
  >
    <img
      alt="poster for Wipeout"
      className="Poster"
      src="https://images2.9c9media.com/image_asset/2021_5_12_39a00768-ae3d-4072-90fa-57a32c91574c_png_2000x3000.jpg"
    />
    <div
      className="Title"
    >
      Wipeout
    </div>
    <button
      className="Btn Add"
      data-testid="WatchlistButton"
      onClick={[Function]}
    >
      (+) Add to Watchlist
    </button>
  </div>
  <div
    className="GridItem"
  >
    <img
      alt="poster for Friends: The Reunion"
      className="Poster"
      src="https://images2.9c9media.com/image_asset/2021_12_8_22ed44f3-d115-4147-9a09-ce75933a0a5a_png_2000x3000.jpg"
    />
    <div
      className="Title"
    >
      Friends: The Reunion
    </div>
    <button
      className="Btn Add"
      data-testid="WatchlistButton"
      onClick={[Function]}
    >
      (+) Add to Watchlist
    </button>
  </div>
  <div
    className="GridItem"
  >
    <img
      alt="poster for The Big Bang Theory"
      className="Poster"
      src="https://images2.9c9media.com/image_asset/2021_2_10_e51a1e22-cce2-4882-9a01-39c248323671_png_2000x3000.jpg"
    />
    <div
      className="Title"
    >
      The Big Bang Theory
    </div>
    <button
      className="Btn Add"
      data-testid="WatchlistButton"
      onClick={[Function]}
    >
      (+) Add to Watchlist
    </button>
  </div>
  <div
    className="GridItem"
  >
    <img
      alt="poster for Pillow Talk"
      className="Poster"
      src="https://images2.9c9media.com/image_asset/2022_2_7_63cfd872-e2c6-4e05-8ec4-2277d76d00df_png_2000x3000.jpg"
    />
    <div
      className="Title"
    >
      Pillow Talk
    </div>
    <button
      className="Btn Add"
      data-testid="WatchlistButton"
      onClick={[Function]}
    >
      (+) Add to Watchlist
    </button>
  </div>
  <div
    className="GridItem"
  >
    <img
      alt="poster for Free Guy"
      className="Poster"
      src="https://images2.9c9media.com/image_asset/2022_2_2_6f62354b-1656-44f8-8713-43e588615890_png_2000x3000.jpg"
    />
    <div
      className="Title"
    >
      Free Guy
    </div>
    <button
      className="Btn Add"
      data-testid="WatchlistButton"
      onClick={[Function]}
    >
      (+) Add to Watchlist
    </button>
  </div>
  <div
    className="GridItem"
  >
    <img
      alt="poster for A Black Lady Sketch Show"
      className="Poster"
      src="https://images2.9c9media.com/image_asset/2021_8_26_9eb5c57f-d21a-4281-b830-6409a590de84_png_2000x3000.jpg"
    />
    <div
      className="Title"
    >
      A Black Lady Sketch Show
    </div>
    <button
      className="Btn Add"
      data-testid="WatchlistButton"
      onClick={[Function]}
    >
      (+) Add to Watchlist
    </button>
  </div>
  <div
    className="GridItem"
  >
    <img
      alt="poster for Dream Horse"
      className="Poster"
      src="https://images2.9c9media.com/image_asset/2022_2_11_ec4a34d9-98fb-4eca-bd4e-20e0ebfaa8b6_png_2000x3000.jpg"
    />
    <div
      className="Title"
    >
      Dream Horse
    </div>
    <button
      className="Btn Add"
      data-testid="WatchlistButton"
      onClick={[Function]}
    >
      (+) Add to Watchlist
    </button>
  </div>
  <div
    className="GridItem"
  >
    <img
      alt="poster for Bridget Everett: Gynecological Wonder"
      className="Poster"
      src="https://images2.9c9media.com/image_asset/2022_2_4_31e891c5-e014-4898-ac41-8be029044360_png_2000x3000.jpg"
    />
    <div
      className="Title"
    >
      Bridget Everett: Gynecological Wonder
    </div>
    <button
      className="Btn Add"
      data-testid="WatchlistButton"
      onClick={[Function]}
    >
      (+) Add to Watchlist
    </button>
  </div>
  <div
    className="GridItem"
  >
    <img
      alt="poster for RuPaul's Drag Race: UK vs The World"
      className="Poster"
      src="https://images2.9c9media.com/image_asset/2022_1_31_fe6f0609-8f89-4811-917b-488584f7daf9_png_2000x3000.jpg"
    />
    <div
      className="Title"
    >
      RuPaul's Drag Race: UK vs The World
    </div>
    <button
      className="Btn Add"
      data-testid="WatchlistButton"
      onClick={[Function]}
    >
      (+) Add to Watchlist
    </button>
  </div>
  <div
    className="GridItem"
  >
    <img
      alt="poster for The Bernie Mac Show"
      className="Poster"
      src="https://images2.9c9media.com/image_asset/2020_10_27_4d1bcf2f-5aa8-4900-aab0-81695c7e0642_png_2000x3000.jpg"
    />
    <div
      className="Title"
    >
      The Bernie Mac Show
    </div>
    <button
      className="Btn Add"
      data-testid="WatchlistButton"
      onClick={[Function]}
    >
      (+) Add to Watchlist
    </button>
  </div>
  <div
    className="GridItem"
  >
    <img
      alt="poster for Real Time With Bill Maher"
      className="Poster"
      src="https://images2.9c9media.com/image_asset/2021_8_26_2c247d30-2a92-4bea-9d31-6ddb5c628063_png_2000x3000.jpg"
    />
    <div
      className="Title"
    >
      Real Time With Bill Maher
    </div>
    <button
      className="Btn Add"
      data-testid="WatchlistButton"
      onClick={[Function]}
    >
      (+) Add to Watchlist
    </button>
  </div>
</section>
`;
